### PR TITLE
Fix importation of compressed points to check for invalid subgroup, i…

### PR DESCRIPTION
**Pull request by dbmcclain**
…n addition to small subgroup attack, and not on curve attack.

### CODE REVIEW

**Comment by dbmcclain** [comment]
The test for subgroup membership was particularly costly, as we have to multiply by the Fr modulus (a huge ragged integer). And then the results of the multiplication are tested and discarded. This happened on every point decompression - which is every time an ECC point (public key) was deserialized.

So rather than doing that, we now pre-scale all points by 1/4 on the way to compressed form, and re-inflate decompressed points by 4 using two doublings (no expensive multiply). This also means that, except for the point at infinity in a compressed form, and points that specifically don't lie on the curve, all other encodings are valid and will automatically decompress into the principal subgroup.

File: crypto/src/curve1174/cpt.rs [hunk 1]
```diff
@@ -79,7 +79,8 @@ impl Pt {
     /// Create from an uncompressed point.
     #[inline]
     pub fn compress(ept: ECp) -> Pt {
-        let bytes = ECp::to_bytes(&ept);
+        let pt_4 = ept / Fr::from(4);
```
Comment (rtsisyk) [inline crypto/src/curve1174/cpt.rs]: What is it? Why it changes genesis block?

Comment (dbmcclain) [reply crypto/src/curve1174/cpt.rs]: The compressed points now look like they belong to ECC points that are scaled by 1/4 from what they were before. This factor gets reinflated on point decompression using fast ECC arithmetic instead of scalar multiply.

TL;DR - the encoding of all Public Keys changes, even though they represent the same effective values as before.

Comment (dbmcclain) [reply crypto/src/curve1174/cpt.rs]: I just provided special purpose (fast) ECC point rescaling by Fr(1/4). No more constant overhead of conversion from int64 to Fr, then Fr::invert(), then conversion to WindowVector. Just uses a constant WindowVector for that scale factor and does a fast raw ECC multiply.

Comment (dbmcclain) [reply crypto/src/curve1174/cpt.rs]: See my previous note about the extreme overhead in our previous point compression / decompression

**Comment by dbmcclain** [comment]
eh? Can't rebase with binary formats? How to fix these conflicts?

**Comment by rtsisyk** [comment]
Please rebase.

**Comment by rtsisyk** [comment]
MErged
